doc/tisplit.1
.\" $Id: tisplit.1,v 1.6 2024/07/12 22:56:16 tom Exp $
.TH TISPLIT 1 2024-07-12 MiscTools "User commands"
.ie \n(.g \{\
.ds `` \(lq
.ds '' \(rq
.\}
.el \{\
.ie t .ds `` ``
.el   .ds `` ""
.ie t .ds '' ''
.el   .ds '' ""
.\}
.hy 0
.SH NAME
tisplit \-
split a terminfo source-file
.SH SYNOPSIS
.B tisplit
[
.I options
] [
.I filename
[
.I ...
]
]
.
.SH DESCRIPTION
Splits a terminfo file (usually terminfo.src) into
a parent file which can be reprocessed by timerge, and
the entries in a \*(``TiSplit\*('' subdirectory.
.PP
If no filename is given, this splits \*(``terminfo.src\*(''
.PP
While splitting, \fBtisplit\fP creates a map of the input file,
named for the input with \*(``.in\*('' appended,
e.g., \*(``terminfo.src.in\*(''
(used by \fBtimerge\fP when reconstructing the original file).
.
.SH OPTIONS
.
.TP 5
.B \-v
verbose, printing the filenames which are split.
.
.
.SH BUGS
.
Probably.
.
.SH EXAMPLES
.
\fBTisplit\fR is used to compare sources for large terminfo databases.
.
.SH SEE ALSO
timerge


CHANGES
2025-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/sue/sue.c: workaround for broken NetBSD header

	* td_getopt.h: NetBSD lint

	* package/debian/changelog, package/misc_tools.spec: bump

2025-09-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* COPYING.asc: sign

	* package/misc_tools.spec: SPDX equivalent

	* COPYING: consolidate license

	* install-bin.in:
	amend $INSTALL in case install-sh is used from a subdirectory

	* install-bin.in: shellcheck warnings

	* config_h.in, doc/minstall.in, package/debian/postinst,
	  src/newpath/makefile.wnt, configure.in:
	trim-blanks

	* src/unmap/chrcount.c, src/unmap/map.c, src/unmap/map_s.c,
	  src/unmap/unmap.c, src/unmap/unmap_s.c, src/width/width.c,
	  src/splitit/splitit.c, src/perror/perror.c, src/slowcat/slowcat.c,
	  src/sue/sue.c, src/tisplit/timerge.c, src/tisplit/tisplit.c,
	  src/acsplit/acmerge.c, src/count_files/count_files.c, src/cpd/cpd.c,
	  src/newest/newest.c, src/newpath/newpath.c, src/acsplit/acsplit.c,
	  td_getline.h:
	s/0/NULL/

	* doc/newpath.1: bump

	* doc/newpath.1: grammar

	* doc/cpd.1: bump

	* doc/cpd.1: grammar

	* doc/chrcount.1: bump

	* doc/chrcount.1: grammar

	* doc/acsplit.1: bump

	* doc/acsplit.1: grammar

	* package/debian/copyright: bump

	* configure: regen

	* package/debian/changelog, package/misc_tools.spec: bump

2025-08-05  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2025-06-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: resync with my-autoconf

2024-07-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: update to 2024-07-27

2024-07-15  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-bin.in: source parameter may have $DESTDIR prefix

	* src/sue/use-usual:
	use $DestDir rather than $DESTDIR, and use the install program
	to install the prepared wrapper rather than fixing after install.

	* package/misc_tools.spec: use --with-execdir option

	* src/sue/use-hacks: quoting per shellcheck warnings
	do not exit on the first error, to work with rpmbuild

	* package/debian/changelog: bump

2024-07-13  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/newpath/newpath.c:
	a relative link for newpath, i.e., making /usr/bin/newpath a symbolic link
	to ../libexec/misctools/newpath did not work -- fix this by finding the
	program on the path, and using realpath to resolve links in that.

	* install-bin.in: corrected link source

	* package/debian/postinst: using libexec now, rather than lib

	* src/sue/use-hacks:
	s/DESTDIR/DestDir/ to avoid conflict with $DESTDIR used in package builds

	* configure: regen

	* src/sue/use-hacks: exit on the first error

	* install-bin.in: add special case for src/sue

	* configure.in: add install_bin symbol, for special case in src/sue

	* src/sue/makefile.in:
	bin-directory depends upon the --with-execdir option

	* configure: regen

	* install-bin.in: account for DESTDIR

	* configure.in: improve message-result for logging

	* package/debian/postinst:
	symbolic link is now done via configure option --with-execdir

	* package/debian/rules: use --with-execdir option

	* configure: regen

	* aclocal.m4: add CF_PROG_LN_S

	* install-bin.in: RCS_BASE

	* package/misc_tools.spec:
	simplify by adding a configure option --with-execdir

	* configure.in, src/acsplit/makefile.in, src/count_files/makefile.in,
	  src/cpd/makefile.in, src/hex/makefile.in, src/isatty/makefile.in,
	  src/newest/makefile.in, src/newpath/makefile.in,
	  src/perror/makefile.in, src/realpath/makefile.in,
	  src/slowcat/makefile.in, src/splitit/makefile.in, src/sue/makefile.in,
	  src/tisplit/makefile.in, src/unmap/makefile.in, src/width/makefile.in:
	added --with-execdir option to simplify installing the tools into libexecdir
	while making a symbolic link to newpath

	* makefile.in: provide for cleanup of install-bin script

	* package/misc_tools.spec:
	build all programs, since there is no naming conflict now

	* doc/minstall.in: update to match unquoted section number in .TH line

	* package/debian/changelog, package/misc_tools.spec: bump

2024-07-12  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/misc_tools.spec:
	add relative symlink for newpath to the %files list

	* doc/newpath.1:
	explain how to reduce environment conflicts with this set of utilities

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/perror.1, doc/realpath.1, doc/slowcat.1, doc/splitit.1, doc/sue.1,
	  doc/timerge.1, doc/tisplit.1, doc/unmap.1, doc/width.1:
	quoting

	* doc/newpath.1: identify source (in lower-left) as "MiscTools"

	* doc/acmerge.1, doc/acsplit.1, doc/chrcount.1, doc/count_files.1,
	  doc/cpd.1, doc/hex.1, doc/isatty.1, doc/map.1, doc/newest.1,
	  doc/newpath.1, doc/perror.1, doc/realpath.1, doc/slowcat.1,
	  doc/splitit.1, doc/sue.1, doc/timerge.1, doc/tisplit.1, doc/unmap.1,
	  doc/width.1:
	check-manpage

	* package/debian/copyright: bump

	* configure: regen

	* COPYING.asc: sign

	* COPYING, package/debian/changelog, package/misc_tools.spec: bump

2024-06-03  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: update to 2024-05-27

2023-12-27  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.sub: 2023/12/27

2023-12-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2023/12/2

2023-12-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* aclocal.m4: resync with my-autoconf

2023-04-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* doc/minstall.in: reduce filename-conflicts by installing ".1mt" files

	* configure: regen


src/tisplit/tisplit.c
/*
 * $Id: tisplit.c,v 1.5 2025/09/10 20:40:21 tom Exp $
 *
 * Title:	tisplit.c - split terminfo.src
 * Author:	T.E.Dickey
 * Created:	16 May 1998
 * Function:	Splits a terminfo file (usually terminfo.src) into
 *		a parent file which can be reprocessed by timerge, and
 *		the entries in a subdirectory.
 */
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>

#include <td_getline.h>

#define VERBOSE(n) if(verbose > n) printf
#define isname(c) (isprint((unsigned char)c) && (c) != '#' && !isspace((unsigned char)c) && (c) != '|')

static const char *target = "TiSplit";
static int verbose;

static void
failed(const char *msg)
{
    perror(msg);
    exit(EXIT_FAILURE);
}

/* parse the terminal name from the first line of the description */
static int
defined(char *line, char *name)
{
    if (isname(*line)) {
	char *s = line;
	char *t = name;
	while (isname(*s))
	    *t++ = *s++;
	*t = 0;
	return 1;
    }
    return 0;
}

static char *
my_temp(char *dst)
{
    sprintf(dst, "%s/%s.out", target, target);
    return dst;
}

static FILE *
finish(FILE *ofp, char *name)
{
    char temp[BUFSIZ];

    my_temp(temp);
    if (ofp != NULL)
	fclose(ofp);

    if (*name != 0) {
	char dst[BUFSIZ];

	VERBOSE(1) ("finish %s\n", name);

	sprintf(dst, "%s/%s", target, name);
	rename(temp, dst);
	*name = 0;

	remove(temp);
    }
    if ((ofp = fopen(temp, "w")) == NULL)
	failed(temp);
    return ofp;
}

static FILE *
append(FILE *ofp, FILE *hdr)
{
    char temp[BUFSIZ];

    if (ofp != NULL) {
	char *bfr = NULL;
	size_t have = 0;

	fclose(ofp);
	my_temp(temp);
	VERBOSE(1) ("append %s\n", temp);

	ofp = fopen(temp, "r");
	while (getline(&bfr, &have, ofp) >= 0) {
	    VERBOSE(2) ("*...%s", bfr);
	    fputs(bfr, hdr);
	}
	fclose(ofp);
	free(bfr);

	remove(temp);
	if ((ofp = fopen(temp, "w")) == NULL)
	    failed(temp);
    }
    return ofp;
}

static void
tisplit(const char *path)
{
    FILE *hdr;
    FILE *ifp = fopen(path, "r");
    FILE *ofp;
    char name[BUFSIZ];
    char *bfr = NULL;
    size_t have = 0;

    if (ifp == NULL)
	failed(path);
    mkdir(target, 0777);

    sprintf(name, "%s.in", path);
    remove(name);
    if ((hdr = fopen(name, "w")) == NULL)
	failed(name);

    name[0] = 0;
    ofp = finish((FILE *) 0, name);

    while (getline(&bfr, &have, ifp) >= 0) {
	VERBOSE(2) (">...%s", bfr);
	if (*bfr == '#' || *bfr == '\n' || *bfr == '\r') {
	    if (*name != 0) {
		ofp = finish(ofp, name);
	    }
	} else if (isname(*bfr) && (*name != 0)) {
	    ofp = finish(ofp, name);
	}
	fputs(bfr, ofp);
	if (defined(bfr, name)) {
	    fprintf(hdr, "%s\n", name);
	}
    }
    if (*name != 0)
	ofp = finish(ofp, name);
    else
	ofp = append(ofp, hdr);
    fclose(hdr);
    remove(my_temp(name));

    free(bfr);
}

int
main(int argc, char *argv[])
{
    int n;
    int found = 0;

    if (argc > 1) {
	for (n = 1; n < argc; n++) {
	    if (!strcmp(argv[n], "-v"))
		verbose++;
	    else {
		tisplit(argv[n]);
		found++;
	    }
	}
    }
    if (!found)
	tisplit("terminfo.src");

    return EXIT_SUCCESS;
}
